### ISSUE TITLE
Low training metrics

### ISSUE THREAD

**Issue by runnerup96** [RUCKBReasoning/RESDSQL#57]
Hello!

I am trying to reproduce the training procedure of training RESDSQL with T5-base on text2sql task. I took the original `train.json` and `dev.json` SPIDER files from leaderboard (https://yale-lily.github.io/spider). I follow training procedure given in the README file.

I took the default script parameters from `scripts/train/text2sql` folder:
1. I used Roberta Large(https://huggingface.co/roberta-large) for schema item classifier training and T5-base (https://huggingface.co/t5-base) for text2sql training.
2. I reduced the batch size to **8** from **16** while increasing gradient descent step to **4** from **2** in order to fit into my resources.
3. I fixed seed **42** along with other seeds for all train phases.

I trained for 128 epochs and evaluated using this repo evaluation script and reevaluated with https://github.com/taoyds/test-suite-sql-eval

I get **EM 0.489** and **EX 0.527** - while in the paper its stated that this configuration achieves  **EM 0.717** and **EX 0.779** on dev set.
Can you please point me where I could get the training wrong? I am evaluating different text2sql model for OOD robustness and choose your solution as one of the current SoTA in that domain.

**Comment by lihaoyang-ruc** [comment]
Hello,

Is your setup identical to ours? We conduct model training on the Ubuntu operating system, and the specifics of our Python environment are detailed in the README file.

Additionally, since RESDSQL operates in two stages, you have the option to substitute one of the stages with our published checkpoints. This should help determine which part is yielding incorrect results.

**Comment by lihaoyang-ruc** [comment]
Have you conducted evaluations on all the intermediate checkpoints for the T5-base model? Your results appear to indicate that only the first checkpoint was assessed.

**Comment by runnerup96** [comment]
Yes, I've followed  

> Hello,
> 
> Is your setup identical to ours? We conduct model training on the Ubuntu operating system, and the specifics of our Python environment are detailed in the README file.
> 
> Additionally, since RESDSQL operates in two stages, you have the option to substitute one of the stages with our published checkpoints. This should help determine which part is yielding incorrect results.

Yes, I've followed README environment installation steps, so I have the same environment.

At first, I've have used your published checkpoints for inference and did get the same results as in the paper. Its the training metrics that do not match in my case.

**Comment by runnerup96** [comment]
> Have you conducted evaluations on all the intermediate checkpoints for the T5-base model? Your results appear to indicate that only the first checkpoint was assessed.

I've conducted evaluation on the best checkpoint after 128 epochs.

**Comment by lihaoyang-ruc** [comment]
> > Have you conducted evaluations on all the intermediate checkpoints for the T5-base model? Your results appear to indicate that only the first checkpoint was assessed.
> 
> I've conducted evaluation on the best checkpoint after 128 epochs.

What you are saying is that you have evaluated all intermediate checkpoints (there are a total of 85 checkpoints for training 128 epochs) and the performance of the best of them is **EM 0.489 and EX 0.527**?

**Comment by runnerup96** [comment]
Are you using exact these initial checkpoints?

https://huggingface.co/roberta-large
https://huggingface.co/t5-base

**Comment by lihaoyang-ruc** [comment]
> Are you using exact these initial checkpoints?
> 
> https://huggingface.co/roberta-large https://huggingface.co/t5-base

Yes

**Comment by lihaoyang-ruc** [comment]
Can you show your TensorBoard logs for training the schema item classifier and T5-base? Other than that, can you share the evaluation results if it's convenient? Like this:
<img width="1995" alt="image" src="https://github.com/RUCKBReasoning/RESDSQL/assets/45526716/8935e6dc-05dc-45a9-ab4b-e3feafc85da6">

**Comment by runnerup96** [comment]
Sure!

I decided to spot the source of low training error so I took your item classifier checkpoint(text2sql_schema_item_classifier) and started training just text2sql stage.

For space optimization i keep only last 10 checkpoints. 

I attach my training loss pictures from tensorboard and csv with loss history.
<img width="1501" alt="checkpoints" src="https://github.com/RUCKBReasoning/RESDSQL/assets/26788197/27e69afb-33b2-4c72-875f-441b759217df">
<img width="1545" alt="tb_loss" src="https://github.com/RUCKBReasoning/RESDSQL/assets/26788197/277a035e-374e-4166-ba22-a5541a1fbcc4">
<img width="356" alt="tb_lr_rate" src="https://github.com/RUCKBReasoning/RESDSQL/assets/26788197/54a0e624-cfd0-41dc-b923-4c32a368ada0">
[run-.-tag-train loss.csv](https://github.com/RUCKBReasoning/RESDSQL/files/13294231/run-.-tag-train.loss.csv)

**Comment by lihaoyang-ruc** [comment]
Your training losses appear to be correct.

I suspect the issue may lie with the evaluation scripts or the pre-processed dataset.

- Could you delete the evaluation result files (named checkpoints-*.txt) and rerun the `evaluate_text2sql_ckpts.py` script?

- If the performance remains atypical, I recommend using our pre-processed spider-dev dataset for another evaluation attempt: [resdsql_test.json](https://github.com/RUCKBReasoning/RESDSQL/files/13295087/resdsql_test.json)

**Comment by runnerup96** [comment]
Wow, the first suggestion actually worked!

Best EM ckpt: {'ckpt': './models/spider_no_others/text2sql-t5-base_1/checkpoint-101169', 'EM': 0.7156673114119922, 'EXEC': 0.7611218568665378}
Best EXEC ckpt: {'ckpt': './models/spider_no_others/text2sql-t5-base_1/checkpoint-101169', 'EM': 0.7156673114119922, 'EXEC': 0.7611218568665378}
Best EM+EXEC ckpt: {'ckpt': './models/spider_no_others/text2sql-t5-base_1/checkpoint-101169', 'EM': 0.7156673114119922, 'EXEC': 0.7611218568665378}

How come the original training procedure in `scripts/train/text2sql/train_text2sql_t5_base.sh` gave different results?

**Comment by lihaoyang-ruc** [comment]
I don't know the exact reason for your problem, but after I saw your training loss curve, I thought your training procedure was fine (because I used to look at it every day for a while). 

Initial speculation is that there may be a problem with this piece of code in `evaluate_text2sql_ckpts.py`.

<img width="771" alt="image" src="https://github.com/RUCKBReasoning/RESDSQL/assets/45526716/8952a218-1cda-41e9-9ade-c0087ad32b56">

Specifically, I wrote this piece at the time to enable multi-process (multi-GPU) evaluation, and while it worked fine for me at the time, it does have some hidden problems. For example, if you have new checkpoints with the same name as checkpoints that have already been evaluated, our script will skip evaluating the new checkpoints and continue to keep the results of the old checkpoints. 

Thanks for the tip, I'm going to submit a PR to fix this.

**Comment by lihaoyang-ruc** [comment]
This issue is being closed due to a lack of activity. However, if you continue to encounter problems and need further assistance, please don't hesitate to reopen it, and we will be more than happy to help with additional troubleshooting.
